File: pohlite/pohlite.py
#!/usr/bin/python
import sys
if sys.version_info[0]>=3: raw_input=input

M=int(raw_input())
N=int(raw_input())
num=[0]*N
cost=[0]*N
total=0
for j in range(N):
	q,r=[int(e) for e in raw_input().split()]
	total+=q
	num[j]=q
	cost[j]=r
bag=[0]+[-1]*total
for j in range(N):
	for i in reversed(range(num[j],total+1)):
		if bag[i-num[j]]>=0:
			if bag[i]<0 or bag[i]>bag[i-num[j]]+cost[j]:
				bag[i]=bag[i-num[j]]+cost[j]
j=-1
for i in range(M,total+1):
	if j<0 or (bag[i]>=0 and j>bag[i]):
		j=bag[i]
print(j)

File: pohlite/pohlite.scala
//must use spaces rather than tabs, or WA somehow.
object Main extends App{
    val M=readInt()
    val N=readInt()
    val num=new Array[Int](N)
    val cost=new Array[Int](N)
    var total=0
    for(j<-0 to N-1){
        val Array(q,r) = readLine().split(" ").map(_.toInt)
        total+=q
        num(j)=q
        cost(j)=r
    }
    val bag=new Array[Int](total+1)
    for(i<-1 to total)bag(i)=(-1)
    for(j<-0 to N-1){
        for(i<-total to num(j) by -1){
            if(bag(i-num(j))>=0){
                if(bag(i)<0||bag(i)>bag(i-num(j))+cost(j)){
                    bag(i)=bag(i-num(j))+cost(j)
                }
            }
        }
    }
    var j=(-1)
    for(i<-M to total){
        if(j<0||(bag(i)>=0&&j>bag(i)))j=bag(i)
    }
    println(j)
}